### PR TITLE
[processor/cumulativetodelta] Add include/exclude configuration

### LINKED ISSUE #5877: [cumulativetodeltaprocessor] Improve metric name configuration
#4444 originally changed the cumulative to delta processor to default to converting all cumulative metrics to delta. This behavior was reverted and currently the configuration requires a user identify each cumulative metric by exact name for conversion to delta.

One proposal is to implement regex support https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/4444#discussion_r714556594.

Another thought is to do away with the metric name configuration entirely. That is, by installing this processor into your pipeline you're declaring you want delta metrics. This thought is inspired by https://github.com/open-telemetry/opentelemetry-specification/pull/2032 where the ability to configure aggregation temporality at the view level in an SDK was removed. Aggregation temporality would be overridable at the metric reader/exporter.

### PR DESCRIPTION
**Pull request by TylerHelmuth**
**Description:** 

This PR replaces the existing `metrics` configuration option in favor of a more robust `include`/`exclude` system.  The new configuration allows users to specify, either by name or by regex pattern, which metrics should or should not be converted.  If neither include or exclude are supplied, all metrics appropriate metrics will be converted.  If a name matches both an include and exclude rule, exclude will take precedence.

To support existing customers, the `metrics` configuration is still supported.  The `metrics` configuration cannot be used with `include`/`exclude`.  When `metrics` is used, a warning will be logged during startup.

This PR only updates the configuration options used to determine which metrics should be converted.  The underlying convert logic has not been changed.

**Link to tracking Issue:**
Fixes #5877 

**Testing:**
Unit tests were updated/added.  Also tested new configuration locally using hostmetrics receiver.

**Documentation:**
Updated the processor's readme with the new configuration and more examples.  Also clarified that non-monotinic sums are never converted.

### CODE REVIEW

File: processor/cumulativetodeltaprocessor/README.md
Comment (Aneurysm9) [inline processor/cumulativetodeltaprocessor/README.md:2]: Not sure that this change is appropriate for this PR.  Even if we think that this makes the component ready for production use we should handle that separately from this functional change.

File: processor/cumulativetodeltaprocessor/processor.go
Comment (Aneurysm9) [inline processor/cumulativetodeltaprocessor/processor.go:110]:
```suggestion
func (ctdp *cumulativeToDeltaProcessor) shouldConvertMetric(metricName string) bool {
```

This method name should convey that it is a predicate and not a mutator.

File: processor/cumulativetodeltaprocessor/README.md
Comment (TylerHelmuth) [reply processor/cumulativetodeltaprocessor/README.md:2]: Thats fair,  I will add it back.

**Comment by TylerHelmuth** [comment]
@Aneurysm9 it appears the link job is hung, can you re-trigger it?

**Comment by jpkrohling** [comment]
PR updated to address the v0.48.0 release.

**Comment by TylerHelmuth** [comment]
@open-telemetry/collector-contrib-approvers please review.

File: CHANGELOG.md
Comment (mx-psi) [inline CHANGELOG.md:71]:
```suggestion
```

File: processor/cumulativetodeltaprocessor/config.go
Comment (mx-psi) [inline processor/cumulativetodeltaprocessor/config.go:33]: I think this needs to be `Deprecated: ` (with the colon) to be picked up by linters (see [here](https://github.com/golang/go/wiki/Deprecated))
```suggestion
	// Deprecated: List of cumulative metrics to convert to delta.
```

File: CHANGELOG.md
Comment (mx-psi) [inline CHANGELOG.md:22]: We should also have a deprecation note for the `metrics` option

File: processor/cumulativetodeltaprocessor/config.go
Comment (mx-psi) [inline processor/cumulativetodeltaprocessor/config.go:45]: Is this based on the config on a different processor? (Asking since the `metricstransform` processor has [a different approach](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6cb401f1f25d2affcf5a10e737ad1c7e10912206/processor/metricstransformprocessor/config.go#L110) and ideally we want to be as consistent as possible)

File: CHANGELOG.md
Comment (TylerHelmuth) [reply CHANGELOG.md:22]: Want me to add that to this line or make a second entry for the processor in "Deprecations"?

File: processor/cumulativetodeltaprocessor/config.go
Comment (TylerHelmuth) [reply processor/cumulativetodeltaprocessor/config.go:45]: Ya I copied the Include/Exclude approach from the [hostmetrics receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6cb401f1f25d2affcf5a10e737ad1c7e10912206/receiver/hostmetricsreceiver/internal/scraper/diskscraper/config.go#L33).  I used this one bc I felt like have both Include and Exclude provided the most flexibility.

File: CHANGELOG.md
Comment (mx-psi) [reply CHANGELOG.md:22]: I think adding it to 'Deprecations' is cleaner, but up to you :)

**Review by mx-psi — COMMENTED** [review]
Only one last thing and it LGTM (sorry for the slow drip of comments :)): can you add a test for the regexp match type?

**Review by mx-psi — APPROVED** [review]
Yes, I meant on `config_test.go` :). This LGTM, but I will wait for any comments by @Aneurysm9 before marking as ready to merge

**Comment by TylerHelmuth** [comment]
@Aneurysm9 please review

**Review by pmm-sumo — APPROVED** [review]
I like that it now uses filterset, which brings more consistency to filtering data across the processors. There are some minor conflicts to resolve but otherwise it LGTM

**Comment by mx-psi** [comment]
I will wait a couple more days for Anthony and mark as ready to merge then (ping me if I forget)

**Review by codeboten — APPROVED** [review]

**Pull request merged**
